nomad/stream/event_publisher.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	if cfg.Logger == nil {
+		cfg.Logger = hclog.NewNullLogger()
+	}
+
+	// Set the event buffer size to a minimum
+	if cfg.EventBufferSize == 0 {
+		cfg.EventBufferSize = 100
+	}
+
+	buffer := newEventBuffer(cfg.EventBufferSize, cfg.EventBufferTTL, cfg.OnEvict)
+	e := &EventPublisher{
+		logger:    cfg.Logger.Named("event_publisher"),
+		eventBuf:  buffer,
+		publishCh: make(chan *structs.Events, 64),
+		subscriptions: &subscriptions{
+			byToken: make(map[string]map[*SubscribeRequest]*Subscription),
+		},
+	}
+
+	go e.handleUpdates(ctx)
+
+	return e
+}
+
+// Returns the current length of the event buffer
+func (e *EventPublisher) Len() int {
+	return e.eventBuf.Len()
+}
+
+// Publish events to all subscribers of the event Topic.
+func (e *EventPublisher) Publish(events *structs.Events) {
+	if len(events.Events) > 0 {
+		e.publishCh <- events
+	}
+}
+
+// Subscribe returns a new Subscription for a given request. A Subscription
+// will receive an initial empty currentItem value which points to the first item
+// in the buffer. This allows the new subscription to call Next() without first checking
+// for the current Item.
+//
+// A Subscription will start at the requested index, or as close as possible to
+// the requested index if it is no longer in the buffer. If StartExactlyAtIndex is
+// set and the index is no longer in the buffer or not yet in the buffer an error
+// will be returned.
+//
+// When a caller is finished with the subscription it must call Subscription.Unsubscribe
+// to free ACL tracking resources. TODO(drew) ACL tracking
+func (e *EventPublisher) Subscribe(req *SubscribeRequest) (*Subscription, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	var head *bufferItem
+	var offset int
+	if req.Index != 0 {
+		head, offset = e.eventBuf.StartAtClosest(req.Index)
+	} else {
+		head = e.eventBuf.Head()
+	}
+	if offset > 0 && req.StartExactlyAtIndex {
+		return nil, fmt.Errorf("requested index not in buffer")
+	} else {
+		e.logger.Debug("requested index no longer in buffer", "requsted", int(req.Index), "closest", int(head.Events.Index))
+	}
+
+	// Empty head so that calling Next on sub
+	start := newBufferItem(&structs.Events{Index: req.Index})
+	start.link.next.Store(head)
+	close(start.link.ch)
+
+	sub := newSubscription(req, start, e.subscriptions.unsubscribe(req))
+
+	e.subscriptions.add(req, sub)
+	return sub, nil
+}
+
+// CloseAll closes all subscriptions
+func (e *EventPublisher) CloseAll() {
+	e.subscriptions.closeAll()
+}
+
+func (e *EventPublisher) handleUpdates(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			e.subscriptions.closeAll()
+			return
+		case update := <-e.publishCh:
+			e.sendEvents(update)
+		}
+	}
+}
+
+// sendEvents sends the given events to the publishers event buffer.
+func (e *EventPublisher) sendEvents(update *structs.Events) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	e.eventBuf.Append(update)
+}
+
+func (s *subscriptions) add(req *SubscribeRequest, sub *Subscription) {
```

Comment:
I may suggest moving the subscription type definition below to be closer to its methods - it took me a second to notice the change of method receivers.

Reply:
sorry about that :) will move